controllers/stardoguser_controller.go

```diff
@@ -18,36 +18,170 @@ package controllers
 
 import (
 	"context"
+	"fmt"
+	"github.com/vshn/stardog-userrole-operator/stardogrest"
+	apierrors "k8s.io/apimachinery/pkg/api/errors"
+	"k8s.io/apimachinery/pkg/types"
+	"k8s.io/apimachinery/pkg/util/errors"
+	"time"
 
 	"github.com/go-logr/logr"
+	"k8s.io/api/core/v1"
 	"k8s.io/apimachinery/pkg/runtime"
 	ctrl "sigs.k8s.io/controller-runtime"
 	"sigs.k8s.io/controller-runtime/pkg/client"
 
-	stardogv1alpha1 "github.com/vshn/stardog-userrole-operator/api/v1alpha1"
+	. "github.com/vshn/stardog-userrole-operator/api/v1alpha1"
 )
 
 // StardogUserReconciler reconciles a StardogUser object
 type StardogUserReconciler struct {
 	client.Client
-	Log    logr.Logger
-	Scheme *runtime.Scheme
+	Log               logr.Logger
+	Scheme            *runtime.Scheme
+	ReconcileInterval time.Duration
 }
 
 // +kubebuilder:rbac:groups=stardog.vshn.ch,resources=stardogusers,verbs=get;list;watch;create;update;patch;delete
 // +kubebuilder:rbac:groups=stardog.vshn.ch,resources=stardogusers/status,verbs=get;update;patch
 
 func (r *StardogUserReconciler) Reconcile(req ctrl.Request) (ctrl.Result, error) {
-	_ = context.Background()
 	_ = r.Log.WithValues("stardoguser", req.NamespacedName)
+	namespace := req.NamespacedName
+	ctx := context.Background()
+	stardogUser := &StardogUser{}
 
-	// your logic here
+	err := r.Client.Get(ctx, namespace, stardogUser)
+	if err != nil {
+		if apierrors.IsNotFound(err) {
+			r.Log.Info("StardogUser not found, ignoring reconcile.", "StardogUser", namespace)
+			return ctrl.Result{Requeue: false}, nil
+		}
+		r.Log.Error(err, "Could not retrieve StardogUser.", "StardogUser", namespace)
+		return ctrl.Result{Requeue: true, RequeueAfter: r.ReconcileInterval}, err
+	}
 
-	return ctrl.Result{}, nil
+	rc := &ReconciliationContext{
+		context:       ctx,
+		conditions:    make(map[StardogConditionType]StardogCondition),
+		stardogClient: stardogrest.NewExtendedBaseClient(),
+		namespace:     namespace.Namespace,
+	}
+	sur := &StardogUserReconciliation{
+		reconciliationContext: rc,
+		resource:              stardogUser,
+	}
+
+	if r.validateSpecification(sur.resource.Spec) != nil {
+		rc.SetStatusCondition(createStatusConditionInvalid(err))
+		rc.SetStatusCondition(createStatusConditionReady(false, "Specification cannot be validated"))
+		return ctrl.Result{}, r.updateStatus(sur)
+	}
+	rc.SetStatusIfExisting(StardogInvalid, v1.ConditionFalse)
+
+	if r.syncUser(sur) != nil {
+		rc.SetStatusCondition(createStatusConditionErrored(err))
+		rc.SetStatusCondition(createStatusConditionReady(false, "Synchronization failed"))
+		return ctrl.Result{}, r.updateStatus(sur)
+	}
+
```

Comment:
This controller lacks the finalizer and logic to remove users from the stardog instance again. Intended or forgotten?

Reply:
Yeah will have to add that. Do you think I have to block deleting the role if a role is assigned to a User? That would mean to go what has been implemented in StardogInstance.

Reply:
That probably depends on Stardog. If the API allows deleting the role and it also automatically removes the role from user in the background at the same time, then it's fine I guess

Reply:
I missed a lot of logic in the last review, now I added update/delete to both user and role entities.